lib/ansible/module_utils/oracle/oci_utils.py

```diff
@@ -0,0 +1,2007 @@
+import logging.config
+import os
+import tempfile
+from datetime import datetime
+from operator import eq
+
+import time
+
+try:
+    import yaml
+
+    import oci
+    from oci.constants import HEADER_NEXT_PAGE
+
+    from oci.exceptions import (
+        InvalidConfig,
+        InvalidPrivateKey,
+        MissingPrivateKeyPassphrase,
+        ConfigFileNotFound,
+        ServiceError,
+        MaximumWaitTimeExceeded,
+    )
+    from oci.identity.identity_client import IdentityClient
+    from oci.object_storage.models import CreateBucketDetails
+    from oci.object_storage.models import UpdateBucketDetails
+    from oci.retry import RetryStrategyBuilder
+    from oci.util import to_dict, Sentinel
+
+    HAS_OCI_PY_SDK = True
+except ImportError:
+    HAS_OCI_PY_SDK = False
+
+
+from ansible.module_utils.basic import _load_params
+from ansible.module_utils._text import to_bytes
+from ansible.module_utils.six import iteritems
+
+__version__ = "1.6.0-dev"
+
+MAX_WAIT_TIMEOUT_IN_SECONDS = 1200
+
+# If a resource is in one of these states it would be considered inactive
+DEAD_STATES = [
+    "TERMINATING",
+    "TERMINATED",
+    "FAULTY",
+    "FAILED",
+    "DELETING",
+    "DELETED",
+    "UNKNOWN_ENUM_VALUE",
+    "DETACHING",
+    "DETACHED",
+]
+
+# If a resource is in one of these states it would be considered available
+DEFAULT_READY_STATES = [
+    "AVAILABLE",
+    "ACTIVE",
+    "RUNNING",
+    "PROVISIONED",
+    "ATTACHED",
+    "ASSIGNED",
+    "SUCCEEDED",
+    "PENDING_PROVIDER",
+]
+
+# If a resource is in one of these states, it would be considered deleted
+DEFAULT_TERMINATED_STATES = ["TERMINATED", "DETACHED", "DELETED"]
+
+
+def get_common_arg_spec(supports_create=False, supports_wait=False):
+    """
+    Return the common set of module arguments for all OCI cloud modules.
+    :param supports_create: Variable to decide whether to add options related to idempotency of create operation.
+    :param supports_wait: Variable to decide whether to add options related to waiting for completion.
+    :return: A dict with applicable module options.
+    """
+    # Note: This method is used by most OCI ansible resource modules during initialization. When making changes to this
+    # method, ensure that no `oci` python sdk dependencies are introduced in this method. This ensures that the modules
+    # can check for absence of OCI Python SDK and fail with an appropriate message. Introducing an OCI dependency in
+    # this method would break that error handling logic.
+    common_args = dict(
+        config_file_location=dict(type="str", required=False),
+        config_profile_name=dict(type="str", required=False, default="DEFAULT"),
+        api_user=dict(type="str", required=False),
+        api_user_fingerprint=dict(type="str", required=False, no_log=True),
+        api_user_key_file=dict(type="str", required=False),
+        api_user_key_pass_phrase=dict(type="str", required=False, no_log=True),
+        auth_type=dict(
```

Comment:
Do you need this? 
Could you just see if api_key is set or not?

Reply:
There may be a case when `api_key` is set but still an user wants to authenticate using instance principal by setting `auth_type` to `instance_principal`.